add dr2 truth-match catalog config

yymao opened the pull request:
This PR adds add the truth-match catalog config for Run2.21 DR2 (v1). 

@evevkovacs you may be able to use this catalog for validation. ~but keep in mind that:~
~1. The column names may differ slightly from @fjaviersanchez's version~
~2. This version of truth-match only use nearest neighbor, without magnitude cut. There may be more mismatches especially given that DR2 is shallower.~

Updated Dec 12: The two points I noted above no longer apply with the latest set of commits (which fixed point 1) and with https://github.com/LSSTDESC/DC2-production/pull/409 (which fixed point 2).

evevkovacs commented:
@yymao  OK, I was able to load **dc2_object_run2.2i_dr2_with_truth_match** and get column names including 'redhsift_truth'.  I had to switch to the branch origin/u/yymao/truth-match-dr2 to find it.

Ok, so I assumed this was a composite catalog and checked that it had cModel magnitudes. But when I tried to use it in the descqa test I got an error. Apologies if I haven't understood what I'm supposed to be doing here.

Here's the traceback:
 Exception occurred when running validation `Nz_i_DEEP2_JAN` on catalog `dc2_object_run2.2i_dr2_with_truth_match`. Below are stdout/stderr and traceback:
/global/common/software/lsst/common/miniconda/current/envs/desc/lib/python3.7/site-packages/GCR/base.py:160: UserWarning: redshift_true_galaxy not available; using redshift_truth instead
  warnings.warn('{} not available; using {} instead'.format(quantities[0], q))
Traceback (most recent call last):
  File "/global/u1/k/kovacs/descqa2-local/descqa_dr2/descqarun/master.py", line 348, in run_tests
    test_result = validation_instance.run_on_single_catalog(catalog_instance, catalog, output_dir_this)
  File "/global/u1/k/kovacs/descqa2-local/descqa_dr2/descqa/NumberDensityVersusRedshift.py", line 269, in run_on_single_catalog
    for catalog_data in catalog_instance.get_quantities(required_quantities, filters=self.filters, return_iterator=True):
  File "/global/common/software/lsst/common/miniconda/current/envs/desc/lib/python3.7/site-packages/GCR/base.py", line 463, in _get_quantities_iter
    native_quantity_getter)
  File "/global/common/software/lsst/common/miniconda/current/envs/desc/lib/python3.7/site-packages/GCR/base.py", line 457, in _load_quantities
    native_data = self._obtain_native_data_dict(native_quantities_needed, native_quantity_getter)
  File "/global/common/software/lsst/common/miniconda/current/envs/desc/lib/python3.7/site-packages/GCR/composite.py", line 306, in _obtain_native_data_dict
    if cat.matching_by_column:
AttributeError: 'CompositeSpecs' object has no attribute 'matching_by_column'
[INFO ][2020-12-11 16:43:57,342] All done! Status report:
--------------------------------------------------
Nz_i_DEEP2_JAN
--------------------------------------------------
cosmoDC2_v1.1.4_small                     VALIDATION_TEST_FAILED
dc2_object_run2.2i_dr2_with_truth_match   RUN_VALIDATION_TEST_ERROR
--------------------------------------------------

[INFO ][2020-12-11 16:43:57,397] Web output: https://portal.nersc.gov/cfs/lsst/descqa/v2/?run=2020-12-11_9

yymao commented:
@evevkovacs thanks for the report. I'm fixing this in https://github.com/yymao/generic-catalog-reader/pull/34

Can you tried the following?

First, if you haven't clone yymao/generic-catalog-reader, clone it. Skip this step if you have done it.

```bash
git clone https://github.com/yymao/generic-catalog-reader.git 
```

Second, switch to the new branch:
```bash
cd /path/to/your/generic-catalog-reader
git fetch
git checkout u/yymao/bug-fix 
```

Third, run descqa:
```
./run_master.sh <your regular options> -p /path/to/your/gcr-catalogs /path/to/your/generic-catalog-reader
```

And see if this fixes the issue? Thank you!

yymao commented:
@evevkovacs -- I have added a new catalog config to this branch `dc2_object_run2.2i_dr2_matched_addon` and it should have exactly the same schema as Javi's catalog. I've updated the instruction above.

yymao commented:
@evevkovacs I went ahead to try the test. Here's the test result: https://portal.nersc.gov/cfs/lsst/descqa/v2/?run=2020-12-12_2&test=Nz_i_DEEP2_JAN

I ran this test with the following command: 

```bash
./run_master.sh -t Nz_i_DEEP2_JAN -c cosmoDC2_v1.1.4_small dc2_object_run2.2i_dr2_matched_addon -p ../gcr-catalogs ../generic-catalog-reader
```

where `../gcr-catalogs` points to this branch (`u/yymao/truth-match-dr2`) and `../generic-catalog-reader` points to the current master of https://github.com/yymao/generic-catalog-reader (v0.9.1).

evevkovacs commented:
@yymao Hi Yao, Tried this  and followed steps above, but I got an error when I tried to checkout u/yymao/bug-fix.  The git fetch worked ok, but git told me the branch didn't exist. I did a git status and guessed that you had already pushed the changes to master, so I did a pull and am trying to run the DESCQA test (step 3 above) now.  Test ran to completion with no errors.  Thanks very much!

yymao commented:
Thanks @evevkovacs! 

Also adding @JoanneBogart as a reviewer for the edits to `GCRCatalogs/dc2_truth_match.py` --- the new edits (method `_use_matchdc2_quantity_modifiers`) enable users to use the new matched catalog as if it were the old one.

JoanneBogart commented:
@yymao  The code in _use_matchdc2_quantity_modifiers looks reasonable and, looking also at dc2_matched_table.DC2MatchedTable, I believe it produces the right set of quantities, but I can't really be sure without loading the two catalogs and comparing.  It seems likely this, or something equivalent, has already been done.

yymao commented:
Thanks @JoanneBogart. This addition of `_use_matchdc2_quantity_modifiers` is mainly for running existing DESCQA test. I did run one successfully with it (see [earlier comment](https://github.com/LSSTDESC/gcr-catalogs/pull/513#issuecomment-743767101)). I wonder if @evevkovacs has run more tests on it and can report back.

evevkovacs commented:
@yymao I tried running the color distribution test (which was the third test in the DC2 survey paper) on dc2_object_run2.2i_dr2_matched_addon  to see if I had any issues with the quantity names. (I don't think it was clear from above comment, but I ran the Nz test on the original truth catalog that you made, which required adding 'redshift_true' to the list of possible redshift names).  Test is still running. I'll post a link when it's done.

evevkovacs commented:
[Here it is](https://portal.nersc.gov/cfs/lsst/descqa/v2/?run=2020-12-13_3&test=Color_Dist_SDSS_with_LSST)

evevkovacs reviewed (approved):
I looked over the files. Nothing jumped out at me, but I am not too familiar with the complexities of composite catalogs, so not all of the options and changes  in dc2_truth_match.py were obvious. The other 3 files looked straightforward. I did test the resulting catalogs with DESCQA and everything worked, so I can sign off at that level.
